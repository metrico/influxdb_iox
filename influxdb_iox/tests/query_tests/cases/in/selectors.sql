-- Demonstrate the use of flux "selector" aggregates via SQL
-- IOX_SETUP: AllTypes

-- select * so the tests are more self describing / understandable
SELECT * from m;

---- Note: all the tests below tests float, int, uint, string, and bool types and then a few together

--------
-- FIRST Selector
--------
select selector_first(float_field, time) from m;
select selector_first(int_field, time) from m;
select selector_first(uint_field, time) from m;
select selector_first(string_field, time) from m;
select selector_first(bool_field, time) from m;

-- Test extracting value and time subfields
select selector_first(float_field, time)['value'], selector_first(float_field, time)['time'] from m;
-- Also with subqueries
select f['value'], f['time'] from (select selector_first(float_field, time) as f from m) as sq;
-- Test with additional fields
select selector_first(float_field, time, bool_field, int_field) from m;


--------
-- LAST Selector
--------
select selector_last(float_field, time) from m;
select selector_last(int_field, time) from m;
select selector_last(uint_field, time) from m;
select selector_last(string_field, time) from m;
select selector_last(bool_field, time) from m;

-- Test extracting value and time subfields
select selector_last(float_field, time)['value'], selector_last(float_field, time)['time'] from m;
-- Also with subqueries
select f['value'], f['time'] from (select selector_last(float_field, time) as f from m) as sq;


--------
-- MIN Selector
--------
select selector_min(float_field, time) from m;
select selector_min(int_field, time) from m;
select selector_min(uint_field, time) from m;
select selector_min(string_field, time) from m;
select selector_min(bool_field, time) from m;

-- Test extracting value and time subfields
select selector_min(float_field, time)['value'], selector_min(float_field, time)['time'] from m;
-- Also with subqueries
select f['value'], f['time'] from (select selector_min(float_field, time) as f from m) as sq;


--------
-- MAX Selector
--------
select selector_max(float_field, time) from m;
select selector_max(int_field, time) from m;
select selector_max(uint_field, time) from m;
select selector_max(string_field, time) from m;
select selector_max(bool_field, time) from m;

-- Test extracting value and time subfields
select selector_max(float_field, time)['value'], selector_max(float_field, time)['time'] from m;
-- Also with subqueries
select f['value'], f['time'] from (select selector_max(float_field, time) as f from m) as sq;
