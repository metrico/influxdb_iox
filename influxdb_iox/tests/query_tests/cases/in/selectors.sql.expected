-- Test Setup: AllTypes
-- SQL: SELECT * from m;
+------------+-------------+-----------+--------------+------+--------------------------------+------------+
| bool_field | float_field | int_field | string_field | tag  | time                           | uint_field |
+------------+-------------+-----------+--------------+------+--------------------------------+------------+
| false      | 65.0        | 18.0      | fruz         | row1 | 1970-01-01T00:00:00.000000100Z | 40         |
| false      | 62.0        | 21.0      | ba           | row1 | 1970-01-01T00:00:00.000000200Z | 30         |
| false      | 63.0        | 20.0      | baz          | row1 | 1970-01-01T00:00:00.000000300Z | 35         |
| true       | 64.0        | 19.0      | bar          | row1 | 1970-01-01T00:00:00.000000400Z | 20         |
|            | 64.0        |           |              | row1 | 1970-01-01T00:00:00.000000450Z |            |
| true       | 61.0        | 22.0      | foo          | row1 | 1970-01-01T00:00:00.000000500Z | 25         |
|            |             | 64.0      |              | row1 | 1970-01-01T00:00:00.000000550Z |            |
| true       | 66.0        | 17.0      | faa          | row1 | 1970-01-01T00:00:00.000000600Z | 10         |
+------------+-------------+-----------+--------------+------+--------------------------------+------------+
-- SQL: select selector_first(float_field, time) from m;
+----------------------------------------------------+
| selector_first(m.float_field,m.time)               |
+----------------------------------------------------+
| {value: 65.0, time: 1970-01-01T00:00:00.000000100} |
+----------------------------------------------------+
-- SQL: select selector_first(int_field, time) from m;
+----------------------------------------------------+
| selector_first(m.int_field,m.time)                 |
+----------------------------------------------------+
| {value: 18.0, time: 1970-01-01T00:00:00.000000100} |
+----------------------------------------------------+
-- SQL: select selector_first(uint_field, time) from m;
+--------------------------------------------------+
| selector_first(m.uint_field,m.time)              |
+--------------------------------------------------+
| {value: 40, time: 1970-01-01T00:00:00.000000100} |
+--------------------------------------------------+
-- SQL: select selector_first(string_field, time) from m;
+----------------------------------------------------+
| selector_first(m.string_field,m.time)              |
+----------------------------------------------------+
| {value: fruz, time: 1970-01-01T00:00:00.000000100} |
+----------------------------------------------------+
-- SQL: select selector_first(bool_field, time) from m;
+-----------------------------------------------------+
| selector_first(m.bool_field,m.time)                 |
+-----------------------------------------------------+
| {value: false, time: 1970-01-01T00:00:00.000000100} |
+-----------------------------------------------------+
-- SQL: select selector_first(float_field, time)['value'], selector_first(float_field, time)['time'] from m;
+---------------------------------------------+--------------------------------------------+
| selector_first(m.float_field,m.time)[value] | selector_first(m.float_field,m.time)[time] |
+---------------------------------------------+--------------------------------------------+
| 65.0                                        | 1970-01-01T00:00:00.000000100Z             |
+---------------------------------------------+--------------------------------------------+
-- SQL: select f['value'], f['time'] from (select selector_first(float_field, time) as f from m) as sq;
+-------------+--------------------------------+
| sq.f[value] | sq.f[time]                     |
+-------------+--------------------------------+
| 65.0        | 1970-01-01T00:00:00.000000100Z |
+-------------+--------------------------------+
-- SQL: select selector_first(float_field, time, bool_field, int_field) from m;
+-----------------------------------------------------------------------------------+
| selector_first(m.float_field,m.time,m.bool_field,m.int_field)                     |
+-----------------------------------------------------------------------------------+
| {value: 65.0, time: 1970-01-01T00:00:00.000000100, other_1: false, other_2: 18.0} |
+-----------------------------------------------------------------------------------+
-- SQL: select selector_last(float_field, time) from m;
+----------------------------------------------------+
| selector_last(m.float_field,m.time)                |
+----------------------------------------------------+
| {value: 66.0, time: 1970-01-01T00:00:00.000000600} |
+----------------------------------------------------+
-- SQL: select selector_last(int_field, time) from m;
+----------------------------------------------------+
| selector_last(m.int_field,m.time)                  |
+----------------------------------------------------+
| {value: 17.0, time: 1970-01-01T00:00:00.000000600} |
+----------------------------------------------------+
-- SQL: select selector_last(uint_field, time) from m;
+--------------------------------------------------+
| selector_last(m.uint_field,m.time)               |
+--------------------------------------------------+
| {value: 10, time: 1970-01-01T00:00:00.000000600} |
+--------------------------------------------------+
-- SQL: select selector_last(string_field, time) from m;
+---------------------------------------------------+
| selector_last(m.string_field,m.time)              |
+---------------------------------------------------+
| {value: faa, time: 1970-01-01T00:00:00.000000600} |
+---------------------------------------------------+
-- SQL: select selector_last(bool_field, time) from m;
+----------------------------------------------------+
| selector_last(m.bool_field,m.time)                 |
+----------------------------------------------------+
| {value: true, time: 1970-01-01T00:00:00.000000600} |
+----------------------------------------------------+
-- SQL: select selector_last(float_field, time)['value'], selector_last(float_field, time)['time'] from m;
+--------------------------------------------+-------------------------------------------+
| selector_last(m.float_field,m.time)[value] | selector_last(m.float_field,m.time)[time] |
+--------------------------------------------+-------------------------------------------+
| 66.0                                       | 1970-01-01T00:00:00.000000600Z            |
+--------------------------------------------+-------------------------------------------+
-- SQL: select f['value'], f['time'] from (select selector_last(float_field, time) as f from m) as sq;
+-------------+--------------------------------+
| sq.f[value] | sq.f[time]                     |
+-------------+--------------------------------+
| 66.0        | 1970-01-01T00:00:00.000000600Z |
+-------------+--------------------------------+
-- SQL: select selector_min(float_field, time) from m;
+----------------------------------------------------+
| selector_min(m.float_field,m.time)                 |
+----------------------------------------------------+
| {value: 61.0, time: 1970-01-01T00:00:00.000000500} |
+----------------------------------------------------+
-- SQL: select selector_min(int_field, time) from m;
+----------------------------------------------------+
| selector_min(m.int_field,m.time)                   |
+----------------------------------------------------+
| {value: 17.0, time: 1970-01-01T00:00:00.000000600} |
+----------------------------------------------------+
-- SQL: select selector_min(uint_field, time) from m;
+--------------------------------------------------+
| selector_min(m.uint_field,m.time)                |
+--------------------------------------------------+
| {value: 10, time: 1970-01-01T00:00:00.000000600} |
+--------------------------------------------------+
-- SQL: select selector_min(string_field, time) from m;
+--------------------------------------------------+
| selector_min(m.string_field,m.time)              |
+--------------------------------------------------+
| {value: ba, time: 1970-01-01T00:00:00.000000200} |
+--------------------------------------------------+
-- SQL: select selector_min(bool_field, time) from m;
+-----------------------------------------------------+
| selector_min(m.bool_field,m.time)                   |
+-----------------------------------------------------+
| {value: false, time: 1970-01-01T00:00:00.000000100} |
+-----------------------------------------------------+
-- SQL: select selector_min(float_field, time)['value'], selector_min(float_field, time)['time'] from m;
+-------------------------------------------+------------------------------------------+
| selector_min(m.float_field,m.time)[value] | selector_min(m.float_field,m.time)[time] |
+-------------------------------------------+------------------------------------------+
| 61.0                                      | 1970-01-01T00:00:00.000000500Z           |
+-------------------------------------------+------------------------------------------+
-- SQL: select f['value'], f['time'] from (select selector_min(float_field, time) as f from m) as sq;
+-------------+--------------------------------+
| sq.f[value] | sq.f[time]                     |
+-------------+--------------------------------+
| 61.0        | 1970-01-01T00:00:00.000000500Z |
+-------------+--------------------------------+
-- SQL: select selector_max(float_field, time) from m;
+----------------------------------------------------+
| selector_max(m.float_field,m.time)                 |
+----------------------------------------------------+
| {value: 66.0, time: 1970-01-01T00:00:00.000000600} |
+----------------------------------------------------+
-- SQL: select selector_max(int_field, time) from m;
+----------------------------------------------------+
| selector_max(m.int_field,m.time)                   |
+----------------------------------------------------+
| {value: 64.0, time: 1970-01-01T00:00:00.000000550} |
+----------------------------------------------------+
-- SQL: select selector_max(uint_field, time) from m;
+--------------------------------------------------+
| selector_max(m.uint_field,m.time)                |
+--------------------------------------------------+
| {value: 40, time: 1970-01-01T00:00:00.000000100} |
+--------------------------------------------------+
-- SQL: select selector_max(string_field, time) from m;
+----------------------------------------------------+
| selector_max(m.string_field,m.time)                |
+----------------------------------------------------+
| {value: fruz, time: 1970-01-01T00:00:00.000000100} |
+----------------------------------------------------+
-- SQL: select selector_max(bool_field, time) from m;
+----------------------------------------------------+
| selector_max(m.bool_field,m.time)                  |
+----------------------------------------------------+
| {value: true, time: 1970-01-01T00:00:00.000000400} |
+----------------------------------------------------+
-- SQL: select selector_max(float_field, time)['value'], selector_max(float_field, time)['time'] from m;
+-------------------------------------------+------------------------------------------+
| selector_max(m.float_field,m.time)[value] | selector_max(m.float_field,m.time)[time] |
+-------------------------------------------+------------------------------------------+
| 66.0                                      | 1970-01-01T00:00:00.000000600Z           |
+-------------------------------------------+------------------------------------------+
-- SQL: select f['value'], f['time'] from (select selector_max(float_field, time) as f from m) as sq;
+-------------+--------------------------------+
| sq.f[value] | sq.f[time]                     |
+-------------+--------------------------------+
| 66.0        | 1970-01-01T00:00:00.000000600Z |
+-------------+--------------------------------+